docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,187 @@
+| TransitionTypeReclaimHost |            | TransitionTypeUnbindHost |
+-----------------------------            ----------------------------
+            |                                               |
+            V                                               |
+  ------------------------                                  |
+  | HostStatusReclaiming |                                  |
+  ------------------------                                  |
+                |                                           |
+       ---------+--------------------------                 |
+       |                                  |                 |
+    Success                            Failure              |
+       |                                  |                 |
+       V                                  V                 V
+--------------------------------  ----------------------------------------     
+| HostStatusDiscoveringUnbound |  | HostStatusUnbindingPendingUserAction |
+--------------------------------  ----------------------------------------
+
+```
+
+#### Assumptions
+
+This approach only works if the host is currently part of a working cluster
+which the assisted service can access using the kubeconfig generated during
+install.
+
+There will also need to be enough space on the `/boot` partition to download the
+initrd and kernel, but the agent can check for this. Together the initrd and 
+kernel occupy about 90MB.
+
+### Risks and Mitigations
+
+It's possible that something could go wrong during the boot process. A user 
+that doesn't have access to the host infrastructure wouldn't be able to debug 
+or solve this type of problem. In this case the user is no worse off than they
+currently would be and will have to boot the discovery ISO manually.
+
+A timeout will be implemented when the service expects the host to reregister.
+After this time the host will move to the existing unbinding-pending-user-action
+state.
+
+## Design Details [optional]
+
+### Open Questions
+
+1. Will users want to reclaim individual hosts in an existing cluster or will
+   the whole cluster always be decommissioned at the same time?
+  * This will determine how we target the MachineConfig
+  * Ideally we don't want to be running an agent on nodes we know we don't want
+    to reclaim
+2. How long is the kubeconfig for a particular cluster saved?
+  * It needs to still be around whenever a user wants to reclaim hosts
+
+### UI Impact
+
+A UI would need to be developed for this feature, but it won't be used in the 
+cloud so it doesn't need a UI there.
+
+### Test Plan
+
+Any existing end to end late binding test could be expanded to include
+returning the agents to the infraEnv after a successful install.
+
+## Drawbacks
+
+This will only work in environments where the service has access to the
+kubernetes API on the cluster. This specifically means that it won't be useful
+in the SaaS.
+
+## Alternatives
```

Comment:
Another alternative is to require use of a mapi or capi provider for agents. It could implement the whole workflow when a Machine gets deleted of running a new agent and draining the node. And it would work with the SaaS in theory since the source of truth on whether a node should still be in a cluster is whether its Machine resource has been deleted.

Reply:
I like this idea, but as I mentioned in the other thread I think the service should still be providing whatever information we use to define how the agent runs (ignition, pod spec, whatever else it could be).

Reply:
Yep. I'd think a MachineSet maps to an InfraEnv, and somehow the InfraEnv would provide the info needed to run the agent, though I'm not sure it needs to provide the pod spec necessarily.

To be clear, I don't necessarily think this is a better idea, but it is definitely worth listing as an alternative and at least explaining why we're not going this direction.

Reply:
I added an alternative here for MAPI/CAPI, but I don't think it's something we can do because we also want this to be available for a fully boot-it-yourself user.